feat(validation): move to aurelia-validation 0.12+

doktordirk opened the pull request:
Update to aurelia-validation@^0.12.3 See the changed documenation on validation of usage

doktordirk committed:
feat(validation): move to aurelia-validation 0.12+

BREAKING CHANGE: Update to aurelia-validation@^0.12.3 See the changed documenation on validation of usage

doktordirk commented:
usage preview https://github.com/SpoonX/swan-example-client/pull/30

RWOverdijk commented:
Can we get rid of the old validation entirely

doktordirk commented:
what you mean? we are rid of it (here)

RWOverdijk commented:
Oof... Well, I guess I'll have to take time to review this. Will do so soon. :)

RWOverdijk reviewed (approved):
I can't see anything wrong. I haven't tested this yet, but I assume you have. So, 👍

doktordirk committed:
refactor(validation): use Validator interface and fixed doc typos

doktordirk commented on src/entity-manager.js:
```diff
@@ -150,6 +151,14 @@ export class EntityManager {
+     // Set the validator.
+    if (instance.hasValidation() && !(instance.getValidator())) {
+      let validator = this.container.invoke(StandardValidator);
+      this.container.registerInstance(StandardValidator, validator);
```
fyi: that makes the validator transient.

 whether that's needed is another question

doktordirk replied:
not needed. i fixed that

doktordirk committed:
refactor(validation): use Validator singleton

doktordirk commented:
i think i better move the validator to the metadata. wouldn.t change usage

doktordirk commented:
@RWOverdijk currently every entity has it's validator reference set. since it's always the same for for an entity class, this could be made static or moving the reference to the metadata. what would be the best choice here? DI, static, or metadata?

RWOverdijk commented:
@doktordirk The way it was originally designed was that the validator itself lives on the metadata. The validation instance lives on the entity. That makes most sense.

doktordirk committed:
feat(validation): optionally set Validator class

Validator class moved to metadata and can be set with the validation decorators

bas080 commented on doc/api_entity.md:
```diff
@@ -212,21 +212,55 @@ Check if entity has validation enabled.
+
+---------
+
+## .validate(propertyName?, rules?)
```
Are rules specific to one property or could they count for several properties. In that case you might want rules as the first argument as it allows validating multiple properties with one call of the validate method.

doktordirk replied:
that's just for special cases. you usually would have the rules on the entity. so that's only if you want to either for some reason only want to validate one property or apply different rules. 
as for the order.: that's the one of the (aurelia)validator. hence, it seems better to keep it

bas080 commented on src/entity.js:
```diff
@@ -497,39 +487,34 @@ export class Entity {
 
-    if (!this.__validation) {
-      this.enableValidation();
+    if (!this.__validator) {
```
I do not understand this `if` in my opinion this if can be removed completely

doktordirk replied:
hmm. yes

bas080 commented on src/entity.js:
```diff
@@ -542,6 +527,25 @@ export class Entity {
+    }
+
+    return propertyName ? this.getValidator().validateProperty(this, propertyName, rules)
+                        : this.getValidator().validateObject(this, rules);
```
Will this work if rules is not defined? Rules is an optional argument. Should it not default?

doktordirk replied:
the StandardValidator takes care of that. Question would be custom validator. the abstract Validator class has though
- @param rules Optional. If unspecified, the implementation should lookup the rules for the 
  hence they should also take care of that themself

doktordirk committed:
chore(validation): cleanup

doktordirk commented:
there are two things to discuss
a)

``` js
  validate(propertyName, rules) {
    // entities without validation are to be considered valid
    if (!this.hasValidation()) {
      return Promise.resolve([]);
    }
```

b) validation got partly renamed appropriately to validator. that's not BC naturally and may still inconsistent?

RWOverdijk commented:
@doktordirk We can have BC-breaks. But this is the last one.

RWOverdijk reviewed (approved):
Bueno.

doktordirk commented:
getting this out of the way then 😄

TomFoyster commented on README.md:
```diff
@@ -144,14 +147,22 @@ export class Create {
   create () {
     this.requestInFlight = true;
```
I know this is only your demo code - and it's not actually a change in this PR, but;
Could this flag actually ever be checked against?

doktordirk replied:
not quite sure what you want. can you make an own issue for it?

TomFoyster reviewed (commented):
Untested - but looks good to me
